lib/connection_pool.rb

```diff
@@ -53,22 +53,28 @@ def initialize(options = {}, &block)
   end
 
   def with(options = {})
+    success = false # hoisted
     conn = checkout(options)
     begin
-      yield conn
+      (yield conn).tap do
+        success = true # means the connection wasn't interrupted
+      end
     ensure
-      checkin
+      if success
+        # everything is roses, we can safely check the connection back in
+        checkin
+      else
+        @available.discard!(pop_connection)
```

Comment:
You're right. I'm trying to put a self contained example together now.

The fact that I'm seeing `Mysql2::Error: This connection is in use by: #<Thread:0x007fe86c3263b8 sleep>.` now makes me think that ConnectionPool is not waiting with `@available.discard!(pop_connection)` the same way it did with `checkin` and it's allowing another thread to use a resource too soon.

I was trying to show in the above snippet that I even tried to force a wait when `discard!` is called (because it calls `close` or `disconnect` https://github.com/mperham/connection_pool/blob/efaef8bd75bf1e6fe581eae0e84cba5b75c86120/lib/connection_pool/timed_stack.rb#L130-L133) and I still see the error.